src/biogeophys/BalanceCheckMod.F90

```diff
@@ -630,118 +614,102 @@ subroutine BalanceCheck( bounds, &
+           indexc = patch%column(indexp)
+           indexg = patch%gridcell(indexp)
+
+           write(iulog,*)'WARNING: BalanceCheck: surface flux energy balance error (W/m2)'
+           write(iulog,*)'nstep          = ' ,nstep
+           write(iulog,*)'errseb         = ' ,errseb(indexp)
+
+           if ( errseb_max_val > error_thresh ) then
+              write(iulog,*)'clm model is stopping - error is greater than 1e-5 (W/m2)'
+              write(iulog,*)'sabv           = ' ,sabv(indexp)
+              write(iulog,*)'sabg           = ' ,sabg(indexp), ((1._r8- frac_sno(indexc))*sabg_soil(indexp) + &
+                   frac_sno(indexc)*sabg_snow(indexp)),sabg_chk(indexp)
+              write(iulog,*)'forc_tot      = '  ,forc_solad(indexg,1) + forc_solad(indexg,2) + &
+                   forc_solai(indexg,1) + forc_solai(indexg,2)
+
+              write(iulog,*)'eflx_lwrad_net = ' ,eflx_lwrad_net(indexp)
+              write(iulog,*)'eflx_sh_tot    = ' ,eflx_sh_tot(indexp)
+              write(iulog,*)'eflx_lh_tot    = ' ,eflx_lh_tot(indexp)
+              write(iulog,*)'eflx_soil_grnd = ' ,eflx_soil_grnd(indexp)
+              write(iulog,*)'fsa fsr = '        ,fsa(indexp),    fsr(indexp)
+              write(iulog,*)'fabd fabi = '      ,fabd(indexp,:), fabi(indexp,:)
+              write(iulog,*)'albd albi = '      ,albd(indexp,:), albi(indexp,:)
+              write(iulog,*)'ftii ftdd ftid = ' ,ftii(indexp,:), ftdd(indexp,:),ftid(indexp,:)
+              write(iulog,*)'elai esai = '      ,elai(indexp),   esai(indexp)
+              write(iulog,*)'clm model is stopping'
+              call endrun(decomp_index=indexp, clmlevel=namep, msg=errmsg(sourcefile, __LINE__))
+           end if
+
        end if
 
        ! Soil energy balance check
 
-       found = .false.
-       do c = bounds%begc,bounds%endc
-          if (col%active(c)) then
-             if (abs(errsoi_col(c)) > 1.0e-5_r8 ) then
-                found = .true.
-                indexc = c
-             end if
-          end if
-       end do
-       if ( found ) then
-          write(iulog,*)'WARNING: BalanceCheck: soil balance error (W/m2)'
-          write(iulog,*)'nstep         = ',nstep
-          write(iulog,*)'errsoi_col    = ',errsoi_col(indexc)
-          if (abs(errsoi_col(indexc)) > 1.e-4_r8 .and. (DAnstep > skip_steps) ) then
-             write(iulog,*)'clm model is stopping'
-             call endrun(decomp_index=indexc, clmlevel=namec, msg=errmsg(sourcefile, __LINE__))
-          end if
-       end if
+       errsoi_col_max_val  =  maxval( abs(errsoi_col(bounds%begc:bounds%endc)), mask = (errsoi_col(bounds%begc:bounds%endc) < spval) )
+
+       if (errsoi_col_max_val > 1.0e-5_r8 ) then
+           indexc = maxloc( abs(errsoi_col(bounds%begc:bounds%endc)), 1 , mask = (errsoi_col(bounds%begc:bounds%endc) < spval) ) + bounds%begp -1
```

Comment:
`bounds%begp -1` should be `bounds%begc -1`

I didn't notice any other problems like this in my read-through of the changes, but since I found this one error, can you go back and double-check that all the indexing is correct for all variables - i.e., that you're using begp/endp vs. begc/endc in the correct places?